runtime/platform.c

```diff
@@ -105,6 +105,24 @@ void caml_plat_mutex_free(caml_plat_mutex* m)
   check_err("mutex_free", pthread_mutex_destroy(m));
 }
 
+void caml_plat_mutex_reinit(caml_plat_mutex *m)
+{
+#ifdef DEBUG
+  /* The following logic is needed to let caml_plat_assert_all_locks_unlocked()
+     behave correctly in child processes after a fork operation. */
+  if (caml_plat_try_lock(m)) {
+    /* lock was not held at fork time */
+    caml_plat_unlock(m);
+  } else {
+    /* lock was held at fork time, parent process still holds it, but we
+       don't and need to fix lock count */
+    DEBUG_UNLOCK(m);
```

Comment:
After you `fork`, the state of the lock is unchanged. In the child process, which is a new process and obviously doesn't own this lock, we try to acquire it, which tells us whether the parent process holds the lock or not. This allows the debug counter of held locks to be adjusted correctly. It would be one unit too large if the parent process had the lock held at `fork` time.

Reply:
Regarding the use of `caml_plat_mutex_init`, it's still an UB, and this PR does not change anything about that.

Reply:
Ok